python/ray/data/_internal/execution/operators/hash_shuffle.py

```diff
@@ -451,90 +476,129 @@ def start(self, options: ExecutionOptions) -> None:
         self._aggregator_pool.start()
 
     def _add_input_inner(self, input_bundle: RefBundle, input_index: int) -> None:
+
+        shuffle_metrics = self.get_metrics(0)
+        shuffle_bar = self.sub_progress_bar(0)
+
         # TODO move to base class
-        self._metrics.on_input_queued(input_bundle)
-        try:
-            self._do_add_input_inner(input_bundle, input_index)
-        finally:
-            self._metrics.on_input_dequeued(input_bundle)
+        shuffle_metrics.on_input_received(input_bundle)
+        shuffle_metrics.on_input_queued(input_bundle)
 
-    def _do_add_input_inner(self, input_bundle: RefBundle, input_index: int):
         input_blocks_refs: List[ObjectRef[Block]] = input_bundle.block_refs
         input_blocks_metadata: List[BlockMetadata] = input_bundle.metadata
 
-        for block_ref, block_metadata in zip(input_blocks_refs, input_blocks_metadata):
-            # If operator hasn't propagated schemas (for this sequence) to its
-            # aggregator pool, it will need to do that upon receiving of the
-            # first block
-            should_broadcast_schemas = not self._has_schemas_broadcasted[input_index]
-            input_key_column_names = self._key_column_names[input_index]
-            # Compose shuffling task resource bundle
-            shuffle_task_resource_bundle = {
-                "num_cpus": 1,
-                "memory": self._estimate_shuffling_memory_req(block_metadata),
-            }
-
-            cur_shuffle_task_idx = self._next_shuffle_tasks_idx
-            self._next_shuffle_tasks_idx += 1
-
-            # NOTE: In cases when NO key-columns are provided for hash-partitioning
-            #       to be performed on (legitimate scenario for global aggregations),
-            #       shuffling is essentially reduced to round-robin'ing of the blocks
-            #       among the aggregators
-            override_partition_id = (
-                cur_shuffle_task_idx % self._num_partitions
-                if not input_key_column_names
-                else None
+        def _on_partitioning_done(cur_shuffle_task_idx: int):
```

Comment:
I see you're pulling up this method and unfortunately now GH just shows this whole file as pretty much rewritten because of shifting lines and some other changes laid on top. 

It's totally cool to refactor if you want to make things better but please avoid mixing up refactorings like that and critical changes in semantic. It's making it practically impossible for me to effectively review this change and assert correctness of what we're doing.

Let's extract the refactoring part into PR stacked on top so these could be reviewed separately (and much faster)